File: btree.java
class BTree{

	class Node{
		int[] keys;
		Node[] child;
		int t;
		int nKeys;
		boolean isLeaf;
		Node(int t, boolean isLeaf){
			this.t=t
			this.keys = new int[2*t-1];
			this.child = new Node[2*t];
			this.nKeys= 0;
			this.isLeaf=isLeaf;
		}
	}
	Node root;
	int t;
	Btree(int t){
		this.root = null;
		this.t = t;
	}

	public void traverse(Node curNode){
		int i;
		for(i=0;i<curNode.nKeys;i++){
			if(!curNode.isLeaf){
				this.traverse(curNode.child[i]);
			}
			System.out.print(" "+curNode.keys[i]);
		}
		if(!curNode.isLeaf){
			this.traverse(curNode.child[i]);
		}	
	}

	public boolean search(int key){
		boolean this.search(this.root, key);
	}
	public boolean search(Node curNode, int key){
		int i=0;
		while(i<curNode.nKeys && key>curNode.keys[i]){
			i++;
		}
		if(curNode.keys[i]==k){
			return true;
		}
		if(curNode.isLeaf){
			return false;
		}
		return this.search(curNode.child[i], key);
	} 

	public void insert(int key){
		if(this.root==null){
			this.root = new Node(this.t, true);
			this.root.keys[0] = key;
			this.root.nKeys = 1
		}else{
			if(this.root.nKeys==2*this.t-1){
				Node newNode = new Node(this.t, false);
				newNode.child[0]=this.root;
				this.splitChild(newNode, 0, this.root);
				int i=0;
				if(newNode.keys[0]<key){
					i++;
				}
				this.insertNonFull(newNode.child[i], key);
				this.root = newNode;
			}else{
				this.insertNonFull(this.root, key);
			}
		}
	}
	public void insertNonFull(Node curNode, int key){
		int i = curNode.nKeys;
		if(curNode.isLeaf){
			while(i>0 && curNode.keys[i-1]>key){
				curNode.keys[i]=curNode.keys[i-1];
				i--;
			}
			curNode.keys[i]=key;
			curNode.nKeys+=1;
		}else{
			while(i>0 && curNode.keys[i-1]>key){
				i--;
			}
			if(curNode.child[i].nKeys==2*this.t-1){
				this.splitChild(curNode, i, curNode.child[i]);
				if(curNode.keys[i]<key){
					i++;
				}
			}
			this.insertNonFull(curNode.child[i], key)
		}
	}

	public void splitChild(Node curNode, int i, Node cNode){
		Node newNode = new Node(this.t, cNode.isLeaf);
		newNode.nKeys = this.t-1;
		for(int j=0;j<this.t-1;j++){
			newNode.keys[j]=cNode.keys[j+this.t];
		}
		if(!cNode.isLeaf){
			for(int j=0;j<this.t-1;j++){
				newNode.child[i]=cNode.child[j+this.t];
			}
		}
		cNode.nKeys=this.t-1;
		for(int j=curNode.nKeys;j>=i+1;j--){
			curNode.child[j+1]=curNode.child[j];
		}
		curNode.child[i+1]=newNode;
		for(int j=curNode.nKeys-1;j>=i;j--){
			curNode.keys[j+1]=curNode.keys[j];
		}
		curNode.keys[i]=cNode.keys[this.t-1];
		curNode.nKeys+=1;
	}
}

File: stacks_and_queues/Btree.py
# class BTNode:
	# def __init__(self,t,isLeaf):
		# self.t = t
		# self.isLeaf = isLeaf
		# self.keys = [0]*(2*t-1)
		# self.child = [None]*(2*t)
		# nKeys = 0

class BTreeNode:
	def __init__(self, isLeaf=False):
		self.isLeaf = isLeaf
		self.keys = []
		self.child = []

class BTree:
	def __init__(self, t):
		self.root = BTNode(True)
		self.t = t
	
	def printTree(self, node, level=0):
		print("LEVEL: ",level," ",len(node.keys), end=":")
		for each in node.keys:
			print(each, end=" ")
		print()
		level += 1
		if node.child !=[]:
			for each in node.child:
				self.printTree(node.child, level)
	
	def search(self, key, node=None):
		if node is None:
			return self.search(key, self.root)
		i = 0
		while i<len(node.keys) and k> node.keys[i]:
			i+=1
		
		if i<len(node.keys) and k==node.keys[i]:
			return (node,i)
		elif node.isLeaf:
			return None
		else:
			return self.search(key, node.child[i])
	
	def insert(self, key):
		root = self.root
		if len(root.keys) == (2*self.t-1):
			tmp = BTNode()
			self.root = tmp
			tmp.child.insert(0, root)
			self.splitChild(tmp, 0)
			self.insertNonFull(tmp, k)
		else:
			self.insertNonFull(root, k)
	
	def insertNonFull(self, node, key):
		i = len(node.keys)-1
		if node.isLeaf:
			node.keys.append((None, None))
			
			
	
class Btree:
	def __init__(self, t):
		self.root=None
		self.t = t
	
	def traverse(self, node, n):
		for i in range(n):
			if not node.isLeaf and node.child[i] is not None:
				self.traverse(node.child[i], node.child[i].nKeys)
			print(node.keys[i],end=" ")
		if not node.isLeaf and node.child[i] is not None:
			self.traverse(node.child[i], node.child[i].nkeys)
	
	def serach(self,node ,key):
		i=0
		while i< node.nKeys and key > node.keys[i]:
			i+=1
		if node.keys[i]==key:
			return node
		if isLeaf:
			return None
		return self.search(node.child[i], key)
	
	
	def insert(self, key):
		if self.root is None:
			self.root = BTNode(True)
			self.root.keys[0] = key
			return
		if self.root.nKeys == 2*t-1 :
			newBtreeNode = BTNode(t, False)
			newBtreeNode.child[0] = self.root
		
		else:
			
			
	
	def splitChild(self, i, btreeNode):
		newBtreeNode = BTNode(btreeNode.t, btreeNode.isLeaf)
		newBtreeNode.nKeys = t-1
		
		for j in range(0,t-1):
			newBtreeNode.keys[j] = btreeNode.keys[j+t]
		
		if not btreeNode.isLeaf:
			for j in range(0,t):
				newBtreeNode.child[j] = btreeNode.child[t+j]
		btreeNode.nKeys = t-1
		for j in range(btreeNode.nKeys, i, -1):
			btreeNode.child[j+1] = btreeNode.child[j]
		btreeNode.child[i+1] = newBtreeNode	
		
		